doc/developer-guide/documentation/RPC.en.rst

```diff
@@ -0,0 +1,397 @@
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+
+.. include:: ../../common.defs
+
+.. highlight:: cpp
+.. default-domain:: cpp
+
+.. _developer-doc-rpc:
+
+RPC
+*************
+
+========
+Overview
+========
+
+To communicate between :program:`traffic_manager` and :program:`traffic_server` there is an RPC mechanism in
+``mgmt/``. This is a simple serialization style RPC which runs over sockets.
+
+=======================================================
+Runtime Structure :program:`traffic_ctl` to :program:`traffic_server`
+=======================================================
+
+.. uml::
+   :align: center
+
+   hide empty members
+
+   node "traffic_manager"
+   node "traffic_server"
+
+   [traffic_ctl] <-d-> [traffic_manager] : Remote RPC
+   [traffic_cop] <-r-> [traffic_manager] : Remote RPC
+   [other remote clients] <-u-> [traffic_manager] : Remote RPC
+   [traffic_manager] <-r-> [traffic_server] : Local RPC
+   [traffic_server] <-r-> [plugin]
+
+:program:`traffic_manager` opens a socket to recieve commands from remote clients. :program:`traffic_manager` has a socket connection to :program:`traffic_server` to relay RPC commands.
+
+===========================================================
+Sequence Diagram from :program:`traffic_ctl` to :program:`traffic_server`
+===========================================================
+
+Sequence diagram for a command sent from :program:`traffic_ctl` to when it is recieved by a plugin. 
+
+.. image:: ../../uml/images/RPC-sequence-diagram.svg
+    
+.. note::
+
+    Currently a fire and forget model. traffic_manager sends out an asynchoronous signal without any acknowledgement. It then proceeds to send a response itself. 
+
+=======================
+Remote RPC vs Local RPC 
+=======================
+
+The RPC API for remote clients, such as :program:`traffic_ctl`, :program:`traffic_cop`, etc, is different from the RPC API used between :program:`traffic_manager` and :program:`traffic_server`. 
+
+:program:`traffic_manager` acts like a bridge for remote clients to interact with :program:`traffic_server`. Thus, it is currently impossible to do things like have :program:`traffic_ctl` directly send messages to :program:`traffic_server`. Classes suffixed with "Remote", ie. :file:`CoreAPIRemote.cc`, and classes suffixed with "Local", ie. :file:`NetworkUtilsLocal.cc` are for remote and local clients, respectively. The following sections will note which set of RPC's are relevant.
+
+=======================
+Serialization Mechanism
+=======================
+
+.. class:: MgmtMarshall
+
+   This is the class used to marshall data objects. It provides functions to marshall and unmarshall data. Each data object is associated with a field. Fields are ``MgmtMarshallType``:
+
+.. type:: MgmtMarshallType
```

Comment:
This should all be indented to indicate it's part fo `MgmtMarshall`. Note that for documentation we use 3 spaces for indent (because almost all the time it's after `.. `).

Reply:
indented.